src/box/lua/config/utils/schema.lua

```diff
@@ -613,6 +613,112 @@ end
+    -- The journey is finished. Validate and return the new value.
+    if #ctx.journey == 0 then
+        -- Call validate_impl() directly to don't construct a
+        -- schema object.
+        validate_impl(schema, rhs, ctx)
+        return rhs
+    end
+
+    local requested_field = ctx.journey[1]
+    assert(requested_field ~= nil)
+
+    if is_scalar(schema) then
+        walkthrough_error(ctx, 'Attempt to index a scalar value of type %s ' ..
+            'by field %q', schema.type, requested_field)
+    elseif schema.type == 'record' then
+        walkthrough_enter(ctx, requested_field)
+        local field_def = schema.fields[requested_field]
+        if field_def == nil then
+            walkthrough_error(ctx, 'No such field in the schema')
+        end
+
+        walkthrough_assert_table(ctx, schema, data)
+        local field_value = data[requested_field] or {}
+        table.remove(ctx.journey, 1)
+        data[requested_field] = set_impl(field_def, field_value, rhs, ctx)
+        return data
+    elseif schema.type == 'map' then
+        walkthrough_enter(ctx, requested_field)
+        local field_def = schema.value
+
+        walkthrough_assert_table(ctx, schema, data)
+        local field_value = data[requested_field] or {}
+        table.remove(ctx.journey, 1)
+        data[requested_field] = set_impl(field_def, field_value, rhs, ctx)
+        return data
+    elseif schema.type == 'array' then
+        -- TODO: Support 'foo[1]' and `{'foo', 1}` paths. See the
+        -- normalize_path() function.
+        walkthrough_error(ctx, 'Indexing an array is not supported yet')
+    else
+        assert(false)
+    end
+end
+
+-- Set the given `rhs` value at the given path in the `data`.
+--
+-- Important: `data` is assumed as already validated against the
+-- given schema, but `rhs` is validated by the method before the
+-- assignment.
+--
+-- The function checks the path against the schema: it doesn't
+-- allow to use a non-existing field or index a scalar value.
+--
+-- The path is either array-like table or a string in the dot
+-- notation.
+--
+-- local data = {}
+-- schema:set(data, 'foo.bar', 42)
+-- print(data.foo.bar) -- 42
+--
+-- local data = {}
+-- schema:set(data, {'foo', 'bar'}, 42)
+-- print(data.foo.bar) -- 42
+--
+-- Nuances:
+--
+-- * Array indexing is not supported yet.
+-- * A scalar of the 'any' type can't be indexed, even when it is
+--   a table. It is OK to set the whole value of the 'any'
+--   type.
+function methods.set(self, data, path, rhs)
+    local schema = rawget(self, 'schema')
+
+    -- Detect `data` and `path` misordering for the most common
+    -- scenario: a record schema and a string path.
+    if schema.type == 'record' and type(data) ~= 'table' then
+        return set_usage()
+    end
+
+    if path ~= nil then
+        path = normalize_path(path, set_usage)
+    end
+
+    if path == nil or next(path) == nil then
+        error('schema:set: empty path', 0)
```

Comment:
Does it mean, the root of the schema can't be set via this method? AFAIU, the reason is the way `set_impl` is implemented.

I see this behaviour is tested below, but it would be nice to know the reason.

Reply:
There are two possible APIs for this method:

```lua
local res = s:set(res, path, rhs)
```

```lua
s:set(res, path, rhs)
```

The former allows to set the root, but requires to remember about the assignment of the result back to `res` if `path` may point to the root. Each of the APIs has its cons and pros.

The latter doesn't allow to set the root. At least, not always -- the root should be an existing table. And not easily -- we should handle this case separately by removing all the fields and set all the new fields.

My choice is the latter API w/o ability to set the root node and the choice was more or less arbitrary. It appears to be good enough for my goals, so I left it implemented this way.

Added a bullet into the 'nuances' section of the method description to make it explicit:

```diff
diff --git a/src/box/lua/config/utils/schema.lua b/src/box/lua/config/utils/schema.lua
index 9b253d49e8..02f5e31aba 100644
--- a/src/box/lua/config/utils/schema.lua
+++ b/src/box/lua/config/utils/schema.lua
@@ -687,6 +687,8 @@ end
 --
 -- Nuances:
 --
+-- * A root node (pointed by the empty path) can't be set using
+--   this method.
 -- * Array indexing is not supported yet.
 -- * A scalar of the 'any' type can't be indexed, even when it is
 --   a table. It is OK to set the whole value of the 'any'
```

I left it without the motivation description, because, in fact, there is no partucular motivation and it should be OK to change it if we'll feel the need.